### PR TITLE
Cubemap convolution + environment LUT

### PR DESCRIPTION
**Pull request by SomaZ**
This add's the split sum approximation for image based lighting based on the UE4 paper. I also fixed little typo's in the blend-states.

### CODE REVIEW

File: codemp/rd-rend2/glsl/lightall.glsl [hunk 1]
```diff
@@ -532,17 +533,19 @@ vec3 CalcIBLContribution(
 #if defined(USE_CUBEMAP)
 	// EnvironmentBRDF is supposed to work with gloss. Invert roughness for now, but this needs
 	// to be replaced at some point with preconvoluted cubemap code.
-	vec3 reflectance = EnvironmentBRDF(1.0 - roughness, NE, specular);
+	//vec3 reflectance = EnvironmentBRDF(1.0 - roughness, NE, specular);
```
Comment (xycaleth) [inline codemp/rd-rend2/glsl/lightall.glsl]: Please don't leave commented code here :) We can always fetch it back from git later on if we need it.

File: codemp/rd-rend2/glsl/lightall.glsl [hunk 2]
```diff
@@ -332,6 +332,7 @@ uniform sampler2D u_ShadowMap;
 
 #if defined(USE_CUBEMAP)
 uniform samplerCube u_CubeMap;
+uniform sampler2D	u_EnvBrdfMap;
```
Comment (xycaleth) [inline codemp/rd-rend2/glsl/lightall.glsl]: Replace tab with space

File: codemp/rd-rend2/glsl/prefilterEnvMap.glsl [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+in vec4 attr_Position;
+in vec4 attr_TexCoord0;
+
+out vec2 var_ScreenTex;
+
+void main()
+{
+	gl_Position = attr_Position;
+	var_ScreenTex = attr_TexCoord0.xy;
+	//vec2 screenCoords = gl_Position.xy / gl_Position.w;
```
Comment (xycaleth) [inline codemp/rd-rend2/glsl/prefilterEnvMap.glsl]: Remove commented out code

File: codemp/rd-rend2/glsl/prefilterEnvMap.glsl [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+vec3 PrefilterEnvMap( float Roughness, vec3 R )
+{
+	vec3 N = R;
+	vec3 V = R;
+	vec3 PrefilteredColor = vec3(0.0);
+	float TotalWeight = 0.0;
+	uint NumSamples = uint(1024);
+	for ( uint i = uint(0); i < NumSamples; i++ )
```
Comment (xycaleth) [inline codemp/rd-rend2/glsl/prefilterEnvMap.glsl]: Should be able to replace `uint(0)` with `0u` instead.

File: codemp/rd-rend2/glsl/prefilterEnvMap.glsl [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+	return TangentX * H.x + TangentY * H.y + N * H.z;
+}
+
+vec3 PrefilterEnvMap( float Roughness, vec3 R )
+{
+	vec3 N = R;
+	vec3 V = R;
+	vec3 PrefilteredColor = vec3(0.0);
+	float TotalWeight = 0.0;
+	uint NumSamples = uint(1024);
```
Comment (xycaleth) [inline codemp/rd-rend2/glsl/prefilterEnvMap.glsl]: Should be able to replace `uint(1024)` with `1024u`

File: codemp/rd-rend2/glsl/prefilterEnvMap.glsl [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+			TotalWeight += NoL;
+		}
+	}
+	return PrefilteredColor / TotalWeight;
+}
+
+void main()
+{
+	float cubeFace = u_ViewInfo.x;
+	vec2 vector;
```
Comment (xycaleth) [inline codemp/rd-rend2/glsl/prefilterEnvMap.glsl]: Can write this as

```
vec2 vector = (var_ScreenTex - vec2(0.5)) * 2.0
```

File: codemp/rd-rend2/glsl/prefilterEnvMap.glsl [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+	vector.x = (var_ScreenTex.x - 0.5) * 2.0;
+	vector.y = (var_ScreenTex.y - 0.5) * 2.0;
+	// from http://www.codinglabs.net/article_physically_based_rendering.aspx
+
+	vec3 normal = normalize( vec3(vector.xy, 1) );
+    if(cubeFace==2)
```
Comment (xycaleth) [inline codemp/rd-rend2/glsl/prefilterEnvMap.glsl]: Indentation here is a bit messed up?

File: codemp/rd-rend2/glsl/prefilterEnvMap.glsl [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    else if(cubeFace==3)
+        normal = normalize( vec3(vector.x, -1,  vector.y) );
+    else if(cubeFace==0)
+        normal = normalize( vec3(  1, vector.y,-vector.x) );
+    else if(cubeFace==1)
+        normal = normalize( vec3( -1, vector.y, vector.x) );
+    else if(cubeFace==5)
+        normal = normalize( vec3(-vector.x, vector.y, -1) );
+
+	float roughness = u_ViewInfo.y / u_ViewInfo.z;
```
Comment (xycaleth) [inline codemp/rd-rend2/glsl/prefilterEnvMap.glsl]: You could fill `u_ViewInfo.w` with `u_ViewInfo.y / u_ViewInfo.z` (when setting the uniform) and avoid this division here.

File: codemp/rd-rend2/tr_backend.cpp [hunk 9]
```diff
@@ -1799,6 +1799,79 @@ static const void *RB_RotatePic2 ( const void *data )
...
+		cubeMipSize >>= 1;
+		numMips++;
+	}
+	numMips = MAX(1, numMips - 2);
+
+	FBO_Bind(tr.preFilterEnvMapFbo);
+	GL_BindToTMU(cubemap, TB_CUBEMAP);
+
+	GLSL_BindProgram(&tr.prefilterEnvMapShader);
+
```
Comment (xycaleth) [inline codemp/rd-rend2/tr_backend.cpp]: Move definitions of `width` and `height` to just before the loop?

File: codemp/rd-rend2/tr_backend.cpp [hunk 10]
```diff
@@ -1799,6 +1799,79 @@ static const void *RB_RotatePic2 ( const void *data )
...
+	VectorSet4(quadVerts[1], 1, 1, 0, 1);
+	VectorSet4(quadVerts[2], 1, -1, 0, 1);
+	VectorSet4(quadVerts[3], -1, -1, 0, 1);
+
+	texCoords[0][0] = 0; texCoords[0][1] = 0;
+	texCoords[1][0] = 1; texCoords[1][1] = 0;
+	texCoords[2][0] = 1; texCoords[2][1] = 1;
+	texCoords[3][0] = 0; texCoords[3][1] = 1;
+
+	while (cubeMipSize)
```
Comment (xycaleth) [inline codemp/rd-rend2/tr_backend.cpp]: Move definitions of `cubeMipSize` and `numMips` to just before this `while`?

File: codemp/rd-rend2/tr_backend.cpp [hunk 11]
```diff
@@ -1799,6 +1799,79 @@ static const void *RB_RotatePic2 ( const void *data )
...
+		{
+			vec4_t viewInfo;
+			VectorSet4(viewInfo, cubemapSide, level, numMips, 0.0);
+			GLSL_SetUniformVec4(&tr.prefilterEnvMapShader, UNIFORM_VIEWINFO, viewInfo);
+			RB_InstantQuad2(quadVerts, texCoords);
+			qglCopyTexSubImage2D(GL_TEXTURE_CUBE_MAP_POSITIVE_X + cubemapSide, level, 0, 0, 0, 0, width, height);
+		}
+	}
+
+	GL_SelectTexture(0);
```
Comment (xycaleth) [inline codemp/rd-rend2/tr_backend.cpp]: Do you need this?

File: codemp/rd-rend2/tr_glsl.cpp [hunk 12]
```diff
@@ -374,6 +376,20 @@ static size_t GLSL_GetShaderHeader(
 						fbufWidthScale,
 						fbufHeightScale));
 
+	if (r_cubeMapping->integer)
+	{
+		int cubeMipSize = CUBE_MAP_SIZE; // r_cubemapSize->integer;
```
Comment (xycaleth) [inline codemp/rd-rend2/tr_glsl.cpp]: Remove the comment

File: codemp/rd-rend2/tr_glsl.cpp [hunk 13]
```diff
@@ -374,6 +376,20 @@ static size_t GLSL_GetShaderHeader(
...
+		int numRoughnessMips = 0;
+
+		while (cubeMipSize)
+		{
+			cubeMipSize >>= 1;
+			numRoughnessMips++;
+		}
+		numRoughnessMips = MAX(1, numRoughnessMips - 2);
```
Comment (xycaleth) [inline codemp/rd-rend2/tr_glsl.cpp]: Perhaps move this code to calculate the number of mips into a function? The same code appears in another one of your changes too.

File: codemp/rd-rend2/tr_image.cpp [hunk 14]
```diff
@@ -2870,6 +2870,89 @@ static void R_CreateFogImage( void ) {
 }
 
 /*
+================
+R_CreateEnvBrdfLUT
+from https://github.com/knarkowicz/IntegrateDFG
+================
+*/
+#define	LUT_WIDTH	128
```
Comment (xycaleth) [inline codemp/rd-rend2/tr_image.cpp]: Can you move these into the function, and change them into `static const int`?

```
static const int LUT_WIDTH = 128;
static const int LUT_HEIGHT = 128;
```

File: codemp/rd-rend2/tr_image.cpp [hunk 15]
```diff
@@ -2870,6 +2870,89 @@ static void R_CreateFogImage( void ) {
...
+#define	LUT_HEIGHT	128
+static void R_CreateEnvBrdfLUT(void) {
+	if (!r_cubeMapping->integer)
+		return;
+
+	int		x, y;
+	uint16_t	data[LUT_WIDTH][LUT_HEIGHT][4];
+	int		b;
+
+	float const MATH_PI = 3.14159f;
```
Comment (xycaleth) [inline codemp/rd-rend2/tr_image.cpp]: There should already be an `M_PI` definition that you can use.

File: codemp/rd-rend2/tr_image.cpp [hunk 16]
```diff
@@ -2870,6 +2870,89 @@ static void R_CreateFogImage( void ) {
...
+			bias /= sampleNum;
+
+			data[y][x][0] = FloatToHalf(scale);
+			data[y][x][1] = FloatToHalf(bias);
+			data[y][x][2] = 0.0f;
+			data[y][x][3] = 0.0f;
+		}
+	}
+
+	tr.envBrdfImage = R_CreateImage("*envBrdfLUT", (byte*)data, 128, 128, IMGTYPE_COLORALPHA, IMGFLAG_NO_COMPRESSION | IMGFLAG_CLAMPTOEDGE, GL_RGBA16F);
```
Comment (xycaleth) [inline codemp/rd-rend2/tr_image.cpp]: Replace `128` with `LUT_WIDTH` and `LUT_HEIGHT`.

And maybe use `GL_RG16F` instead? I'm not sure if the code supports that format...so don't bother if it doesn't :)

File: codemp/rd-rend2/tr_image.cpp [hunk 17]
```diff
@@ -2870,6 +2870,89 @@ static void R_CreateFogImage( void ) {
...
+	return;
```
Comment (xycaleth) [inline codemp/rd-rend2/tr_image.cpp]: Don't need this here.

File: codemp/rd-rend2/tr_world.cpp [hunk 18]
```diff
@@ -569,13 +569,13 @@ static void R_RecursiveWorldNode( mnode_t *node, int planeBits, int dlightBits,
 		tr.pc.c_leafs++;
 
 		// add to z buffer bounds
-		tr.viewParms.visBounds[0][0] = std::min(node->mins[0], tr.viewParms.visBounds[0][0]);
-		tr.viewParms.visBounds[0][1] = std::min(node->mins[1], tr.viewParms.visBounds[0][1]);
-		tr.viewParms.visBounds[0][2] = std::min(node->mins[2], tr.viewParms.visBounds[0][2]);
+		tr.viewParms.visBounds[0][0] = MIN(node->mins[0], tr.viewParms.visBounds[0][0]);
```
Comment (xycaleth) [inline codemp/rd-rend2/tr_world.cpp]: Don't need these changes?

File: codemp/rd-rend2/tr_shader.cpp [hunk 19]
```diff
@@ -310,7 +310,7 @@ static int NameToSrcBlendMode( const char *name )
 	{
 		if ( r_ignoreDstAlpha->integer )
 		{
-			return GLS_DSTBLEND_ONE;
+			return GLS_SRCBLEND_ONE;
```
Comment (xycaleth) [inline codemp/rd-rend2/tr_shader.cpp]: Why are these here?

Comment (SomaZ) [inline codemp/rd-rend2/tr_shader.cpp]: It's simply wrong, found them because of Eez. The function should return a scr blend not a dst blend. This simply crashes for example the JK2 main menu.

File: codemp/rd-rend2/tr_world.cpp (hunk 18 above)
Comment (SomaZ) [inline codemp/rd-rend2/tr_world.cpp]: I can't compile the code otherwise, so it's needed for me somehow. I'm lazy to investigate why.

File: codemp/rd-rend2/tr_shader.cpp (hunk 19 above)
Comment (xycaleth) [inline codemp/rd-rend2/tr_shader.cpp]: Ah yeah, I see 👍
